node/src/components/linear_chain.rs

```diff
@@ -286,60 +335,135 @@ where
                 if let Err(err) = fs.verify() {
                     warn!(%block_hash, %public_key, %err, "received invalid finality signature");
                     return Effects::new();
                 }
-                debug!(%block_hash, %public_key, "received new finality signature");
-                let sigs = self
-                    .pending_finality_signatures
-                    .entry(public_key)
-                    .or_default();
-                // Limit the memory we use for storing unknown signatures from each validator.
-                if sigs.len() >= MAX_PENDING_FINALITY_SIGNATURES_PER_VALIDATOR {
-                    warn!(
-                        %block_hash, %public_key,
-                        "received too many finality signatures for unknown blocks"
-                    );
-                    return Effects::new();
-                }
-                // Add the pending signature and request the block from storage.
-                if sigs.insert(block_hash, *fs).is_some() {
-                    return Effects::new(); // Signature was already known.
-                }
                 effect_builder
                     .get_block_from_storage(block_hash)
                     .event(move |maybe_block| {
                         let maybe_box_block = maybe_block.map(Box::new);
-                        Event::GetBlockForFinalitySignaturesResult(block_hash, maybe_box_block)
+                        Event::GetBlockForFinalitySignaturesResult(fs, maybe_box_block)
                     })
             }
-            Event::GetBlockForFinalitySignaturesResult(block_hash, None) => {
-                let signers = self
-                    .pending_finality_signatures
-                    .values()
-                    .filter_map(|sigs| sigs.get(&block_hash))
-                    .map(|fs| format!("{}", fs.public_key))
-                    .collect_vec();
-                if !signers.is_empty() {
-                    // We have signatures for an unknown block. Print log messages.
-                    warn!(
-                       %block_hash, signers = %signers.join(", "),
-                       "received signatures for a block that was not found in storage"
+            Event::GetBlockForFinalitySignaturesResult(fs, maybe_block) => {
+                if let Some(block) = &maybe_block {
+                    assert_eq!(
+                        block.hash(),
+                        &fs.block_hash,
+                        "block loaded from storage should have a matching block hash."
                     )
                 }
-                Effects::new()
+                // Check if validator is bonded in the era in which the block was created in.
+                effect_builder
+                    .is_bonded_validator(fs.era_id, fs.public_key)
+                    .map(|is_bonded| {
+                        if is_bonded {
+                            Ok((maybe_block, fs, is_bonded))
```

Comment:
Why not create the events here directly instead of the `result` call below?
```suggestion
                            Event::IsBonded(maybe_block, fs, is_bonded),
```

Reply:
It won't compile. That's b/c we will return a `Future<Event>` and we need to return an `Effects = Vec<Future<Vec<Event>>`.